pkg/cloudfoundry/ManifestUtils.go

```diff
@@ -0,0 +1,226 @@
+package cloudfoundry
+
+import (
+	"fmt"
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"reflect"
+
+	"github.com/SAP/jenkins-library/pkg/log"
+)
+
+const propApplications = "applications"
+const propBuildpacks = "buildpacks"
+const propBuildpack = "buildpack"
+
+// Manifest ...
+type Manifest struct {
+	self     map[string]interface{}
+	modified bool
+	name     string
+}
+
+var m Manifest
+
+var _readFile = ioutil.ReadFile
+
+// ReadManifest ...
+func ReadManifest(name string) (Manifest, error) {
+
+	log.Entry().Infof("Reading manifest file  '%s'", name)
+
+	m := Manifest{self: make(map[string]interface{}), name: name, modified: false}
+
+	content, err := _readFile(name)
+	if err != nil {
+		return m, errors.Wrapf(err, "cannot read file '%v'", m.name)
+	}
+
+	err = yaml.Unmarshal(content, &m.self)
+
+	if err != nil {
+		return m, errors.Wrapf(err, "Cannot parse yaml file '%s': %s", m.name, string(content))
+	}
+
+	log.Entry().Infof("Manifest file '%s' has been parsed", m.name)
+
+	return m, nil
+}
+
+// WriteManifest ...
+func (m *Manifest) WriteManifest() error {
+
+	d, err := yaml.Marshal(&m.self)
+	if err != nil {
+		return err
+	}
+
+	log.Entry().Debugf("Writing manifest file '%s'", m.name)
+	err = ioutil.WriteFile(m.name, d, 0644)
+
+	if err == nil {
+		m.modified = false
+	}
+
+	log.Entry().Debugf("Manifest file '%s' has been written", m.name)
+	return err
+}
+
+// GetName ...
+func (m Manifest) GetName() string {
+	return m.name
+}
+
+// GetApplications ...
+func (m Manifest) GetApplications() ([]interface{}, error) {
+	return toSlice(m.self)
+}
+
+// ApplicationHasProperty Checks if the application denoted by 'index' has the property 'name'
+func (m Manifest) ApplicationHasProperty(index int, name string) (bool, error) {
+
+	sliced, err := toSlice(m.self[propApplications])
+
+	if err != nil {
+		return false, err
+	}
+
+	if index >= len(sliced) {
+		return false, fmt.Errorf("Index (%d) out of bound. Number of apps: %d", index, len(sliced))
+	}
+
+	_m, err := toMap(sliced[index])
+
+	if err != nil {
+		return false, err
+	}
+
+	_, ok := _m[name]
+
+	return ok, nil
+}
+
+// GetApplicationProperty ...
+func (m Manifest) GetApplicationProperty(index int, name string) (interface{}, error) {
+
+	sliced, err := toSlice(m.self[propApplications])
+
+	if err != nil {
+		return "", err
+	}
+
+	if index >= len(sliced) {
+		return "", fmt.Errorf("Index (%d) out of bound. Number of apps: %d", index, len(sliced))
+	}
+
+	app, err := toMap(sliced[index])
+
+	if err != nil {
+		return "", err
+	}
+
+	if app[name] != nil {
+		return app[name], nil
+	}
+
```

Comment:
This prevents setting a property to the value nil in the YAML. I think you need a different check: 
```suggestion
	value, exists := app[name]
	if exists {
		return value, nil
	}

```

Reply:
>This prevents setting a property to the value nil in the YAML

We do not set values in the yaml here. Here we read only and return what we read.

The difference is:

* With the current approach we return `nil` in case the key `name` is present and associated with value `nil` (which is bogus scenario, there should be a real name, even the empty string as name is bogus). In case it it like that we don't return `nil` but we return with an error stating there is no name. In case there is no key `name` we respond also with an error.
* With the suggested approach we return nil in case the key exists (... again: and is associated with value `nil`.

In my opinion it fits better to raise an error in case there is a the key `name` associated with value `nil`. From that point of view I think we should not follow the suggestion.

Reply:
I think you misunderstood. If the user configured a value resolving to "nil" in his manifest, the code will behave as if instead the parameter was not configured at all. The suggested change would take this into account and distinguish between "is present" and "not present" in the map, ignoring the actual value, which may also be "nil".

Reply:
I guess I understood the details (basically having a `name` key mapped to `nil` vs having no `name` key at all). My point is: in case the user explicitly configured `nil` in the yaml file we will return that `nil` value. I think that does not make sense since `nil` is not a valid value for an application name.

Reply:
ahhh. Maybe too much in parallel. Here we are not talking about the `name` property of the application, but about any property. OK. There might be some properties where `nil` is a reasonable value. Hence it might make sense to return the nil value. When resolving the `name` property for an application we have to check when we call that function here. Now I agree with you ...